File: tutorials/clojure/02-hello-web-server/test/hello/server_test.clj
(ns hello.server-test
  (:require [clojure.test :refer :all]
            [hello.server :refer :all]
            [clojure.data.json :as json]))

(deftest root-route-test
  (let [response (app {:request-method :get :uri "/"})]
    (is (= 200 (:status response)))
    (is (.contains (:body response) "Hello, world!"))))

(deftest greet-route-test
  (let [response (app {:request-method :get :uri "/greet/Alice"})]
    (is (= 200 (:status response)))
    (is (.contains (:body response) "Hello, Alice!"))))

(deftest not-found-test
  (let [response (app {:request-method :get :uri "/unknown"})]
    (is (= 404 (:status response)))
    (is (.contains (:body response) "not found"))))

File: tutorials/golang/02-hello-web-server/server_test.go
package main

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestRoutes(t *testing.T) {
	mux := newMux()

	tests := []struct {
		name       string
		path       string
		wantStatus int
		wantMsg    string
	}{
		{"root returns hello", "/", http.StatusOK, "Hello, World!"},
		{"greet by name", "/greet/Gopher", http.StatusOK, "Hello, Gopher!"},
		{"greet another name", "/greet/Alice", http.StatusOK, "Hello, Alice!"},
		{"unknown path is 404", "/unknown", http.StatusNotFound, "not found"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			req := httptest.NewRequest(http.MethodGet, tt.path, nil)
			rec := httptest.NewRecorder()

			mux.ServeHTTP(rec, req)

			if rec.Code != tt.wantStatus {
				t.Errorf("status = %d, want %d", rec.Code, tt.wantStatus)
			}

			var resp Response
			if err := json.NewDecoder(rec.Body).Decode(&resp); err != nil {
				t.Fatalf("failed to decode JSON: %v", err)
			}
			if resp.Message != tt.wantMsg {
				t.Errorf("message = %q, want %q", resp.Message, tt.wantMsg)
			}

			ct := rec.Header().Get("Content-Type")
			if ct != "application/json" {
				t.Errorf("Content-Type = %q, want application/json", ct)
			}
		})
	}
}
